File: macclean/cleaners/security.py
import sqlite3
from pathlib import Path
import click
from rich.console import Console
from rich.table import Table
from rich.panel import Panel
from macclean.core.utils import run_cmd

console = Console()

_TCC_SERVICES = {
    "kTCCServiceCamera": "Camera",
    "kTCCServiceMicrophone": "Microphone",
    "kTCCServiceScreenCapture": "Screen Recording",
    "kTCCServiceLocation": "Location",
    "kTCCServiceAddressBook": "Contacts",
    "kTCCServiceCalendar": "Calendar",
    "kTCCServicePhotos": "Photos",
    "kTCCServiceSystemPolicyAllFiles": "Full Disk Access",
}


def _check(label: str, good: bool, detail: str = "") -> tuple[str, str, str]:
    status = "[green]OK[/]" if good else "[red]OFF[/]"
    return label, status, detail


def _filevault_status() -> tuple[bool, str]:
    out, _ = run_cmd(["fdesetup", "status"])
    on = "FileVault is On" in out
    return on, out.split("\n")[0].strip()


def _firewall_status() -> tuple[bool, str]:
    out, code = run_cmd([
        "/usr/libexec/ApplicationFirewall/socketfilterfw", "--getglobalstate"
    ])
    on = "enabled" in out.lower()
    return on, "Enabled" if on else "Disabled"


def _sip_status() -> tuple[bool, str]:
    out, _ = run_cmd(["csrutil", "status"])
    on = "enabled" in out.lower()
    return on, out.split("\n")[0].strip()


def _gatekeeper_status() -> tuple[bool, str]:
    out, _ = run_cmd(["spctl", "--status"])
    on = "enabled" in out.lower() or "assessments enabled" in out.lower()
    return on, "Enabled" if on else "Disabled"


def _read_tcc_permissions(home: Path) -> list[tuple[str, str, str]]:
    tcc_db = home / "Library" / "Application Support" / "com.apple.TCC" / "TCC.db"
    rows = []
    if not tcc_db.exists():
        return rows
    try:
        conn = sqlite3.connect(str(tcc_db))
        cur = conn.execute(
            "SELECT service, client, auth_value FROM access WHERE auth_value = 2"
        )
        for service, client, _ in cur.fetchall():
            svc_name = _TCC_SERVICES.get(service, service)
            rows.append((svc_name, client, "Allowed"))
        conn.close()
    except Exception:
        pass
    return sorted(rows)


@click.command()
@click.pass_context
def cmd(ctx):
    """Security health check: FileVault, Firewall, SIP, Gatekeeper, app permissions."""
    home = Path.home()

    # System security checks
    checks = []
    fv_ok, fv_detail = _filevault_status()
    checks.append(_check("FileVault (disk encryption)", fv_ok, fv_detail))
    fw_ok, fw_detail = _firewall_status()
    checks.append(_check("Firewall", fw_ok, fw_detail))
    sip_ok, sip_detail = _sip_status()
    checks.append(_check("System Integrity Protection (SIP)", sip_ok, sip_detail))
    gk_ok, gk_detail = _gatekeeper_status()
    checks.append(_check("Gatekeeper", gk_ok, gk_detail))

    sec_table = Table(show_header=True, show_lines=True)
    sec_table.add_column("Check")
    sec_table.add_column("Status")
    sec_table.add_column("Detail")
    for row in checks:
        sec_table.add_row(*row)
    console.print(Panel(sec_table, title="[bold cyan]System Security[/]"))

    # App permissions from TCC
    perms = _read_tcc_permissions(home)
    if perms:
        perm_table = Table(show_header=True, show_lines=True)
        perm_table.add_column("Permission")
        perm_table.add_column("App / Bundle ID")
        perm_table.add_column("Status")
        for svc, client, status in perms:
            perm_table.add_row(svc, client, f"[green]{status}[/]")
        console.print(Panel(perm_table, title="[bold cyan]App Permissions (TCC)[/]"))
    else:
        console.print("[dim]TCC database not readable without Full Disk Access or sudo.[/]")

File: src/cleaners/security.rs
use anyhow::Result;
use colored::Colorize;
use comfy_table::{Table, presets::UTF8_BORDERS_ONLY};
use crate::core::cmd::run_cmd;

pub fn map_service(service: &str) -> String {
    match service {
        "kTCCServiceCamera"              => "Camera".into(),
        "kTCCServiceMicrophone"          => "Microphone".into(),
        "kTCCServiceScreenCapture"       => "Screen Recording".into(),
        "kTCCServiceLocation"            => "Location".into(),
        "kTCCServiceAddressBook"         => "Contacts".into(),
        "kTCCServiceCalendar"            => "Calendar".into(),
        "kTCCServicePhotos"              => "Photos".into(),
        "kTCCServiceSystemPolicyAllFiles"=> "Full Disk Access".into(),
        "kTCCServiceAccessibility"       => "Accessibility".into(),
        "kTCCServiceReminders"           => "Reminders".into(),
        "kTCCServiceUbiquity"            => "iCloud".into(),
        "kTCCServiceShareKit"            => "Sharing".into(),
        other => other.replace("kTCCService", ""),
    }
}

pub fn run() -> Result<()> {
    println!("\n{}", "[ Security Checks ]".cyan().bold());

    // ── Security checks table ─────────────────────────────────────────────────
    let fv = run_cmd(&["fdesetup", "status"])
        .output
        .contains("FileVault is On");

    let fw = {
        let r = run_cmd(&[
            "/usr/libexec/ApplicationFirewall/socketfilterfw",
            "--getglobalstate",
        ]);
        r.output.to_lowercase().contains("enabled")
    };

    let sip = {
        let r = run_cmd(&["csrutil", "status"]);
        r.output.to_lowercase().contains("enabled")
    };

    let gk = {
        let r = run_cmd(&["spctl", "--status"]);
        let out = r.output.to_lowercase();
        out.contains("assessments enabled") || out.contains("enabled")
    };

    let mut table = Table::new();
    table.load_preset(UTF8_BORDERS_ONLY);
    table.set_header(vec!["Security Check", "Status"]);

    for (name, ok) in &[
        ("FileVault",  fv),
        ("Firewall",   fw),
        ("SIP",        sip),
        ("Gatekeeper", gk),
    ] {
        let status = if *ok {
            "OK".green().to_string()
        } else {
            "OFF".red().to_string()
        };
        table.add_row(vec![name.to_string(), status]);
    }
    println!("{}", table);

    // ── TCC permissions table ─────────────────────────────────────────────────
    println!("\n{}", "[ App Permissions (TCC) ]".cyan().bold());

    let home = dirs::home_dir().unwrap_or_else(|| std::path::PathBuf::from("."));
    let db_path = home
        .join("Library/Application Support/com.apple.TCC/TCC.db");

    if !db_path.exists() {
        println!("  TCC database not accessible (requires Full Disk Access).");
        return Ok(());
    }

    let db_str = db_path.to_string_lossy();
    let out = run_cmd(&[
        "sqlite3",
        "-separator",
        "|",
        &db_str,
        "SELECT service,client,auth_value FROM access WHERE auth_value=2",
    ]);

    if !out.success() || out.output.trim().is_empty() {
        println!("  No granted permissions found (or TCC.db not readable).");
        return Ok(());
    }

    let mut perm_table = Table::new();
    perm_table.load_preset(UTF8_BORDERS_ONLY);
    perm_table.set_header(vec!["Permission", "App", "Status"]);

    for line in out.output.lines() {
        let parts: Vec<&str> = line.splitn(3, '|').collect();
        if parts.len() < 3 {
            continue;
        }
        let service = map_service(parts[0].trim());
        let client  = parts[1].trim();
        perm_table.add_row(vec![
            service.to_string(),
            client.to_string(),
            "Allowed".green().to_string(),
        ]);
    }
    println!("{}", perm_table);

    Ok(())
}
